interfaces/builtin/iio.go

```diff
@@ -0,0 +1,144 @@
+	return "iio"
+}
+
+func (iface *IioInterface) String() string {
+	return iface.Name()
+}
+
+// Pattern to match allowed iio device nodes. It is gonna be used to check the
+// validity of the path attributes in case the udev is not used for
+// identification
+var iioControlDeviceNodePattern = regexp.MustCompile("^/dev/iio:device[0-9]+$")
+
+// Check validity of the defined slot
+func (iface *IioInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget snap
+	if !(slot.Snap.Type == "gadget" || slot.Snap.Type == "os") {
+		return fmt.Errorf("%s slots only allowed on gadget or core snaps", iface.Name())
+	}
+
+	// Validate the path
+	path, ok := slot.Attrs["path"].(string)
+	if !ok || path == "" {
+		return fmt.Errorf("%s slot must have a path attribute", iface.Name())
+	}
+
+	path = filepath.Clean(path)
+
+	if !iioControlDeviceNodePattern.MatchString(path) {
+		return fmt.Errorf("%s path attribute must be a valid device node", iface.Name())
+	}
+
+	return nil
+}
+
+// Checks and possibly modifies a plug
+func (iface *IioInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+	// Currently nothing is checked on the plug side
+	return nil
+}
+
+// Returns snippet granted on install
+func (iface *IioInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+// Getter for the security snippet specific to the plug
+func (iface *IioInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	path, pathOk := slot.Attrs["path"].(string)
+	if !pathOk {
+		return nil, nil
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		cleanedPath := filepath.Clean(path)
+		snippet := bytes.Replace(iioConnectedPlugAppArmor, []byte("###IIO_DEVICE_PATH###"), []byte(cleanedPath), -1)
+
+		// The path is already verified against a regular expression
+		// in SanitizeSlot so we can rely on its structure here and
+		// safely strip the '/dev/' prefix to get the actual name of
+		// the IIO device.
+		deviceName := strings.TrimPrefix(path, "/dev/")
+		snippet = bytes.Replace(snippet, []byte("###IIO_DEVICE_NAME###"), []byte(deviceName), -1)
+
+		return snippet, nil
+
+	case interfaces.SecurityUDev:
+		var tagSnippet bytes.Buffer
+		const pathPrefix = "/dev/"
+		const udevRule string = `KERNEL=="%s", TAG+="%s"`
+		for appName := range plug.Apps {
+			tag := udevSnapSecurityName(plug.Snap.Name(), appName)
+			tagSnippet.WriteString(fmt.Sprintf(udevRule, strings.TrimPrefix(path, pathPrefix), tag))
+			tagSnippet.WriteString("\n")
+		}
+		return tagSnippet.Bytes(), nil
+	}
```

Comment:
@jocave we require the udev tagging since if a snap plugged something with udev tagging and something without udev tagging, then the snap would only be allowed access to the devices that were udev tagged since those would be the only devices in the device cgroup. The apparmor rule is so that the access is allowed by apparmor even if the cgroup allows it.

In the case of i2c and iio where the device names are static and won't change, the apparmor rule is technically enough to handle access, but we do both in case the snap is also plugging a slot that is dynamically tagged.